### PR TITLE
Set AWS container properties from container options

### PR DESCRIPTION
**Pull request by manuelesimi**
This contribution addresses the requirements from https://github.com/nextflow-io/nextflow/issues/2282 and https://github.com/nextflow-io/nextflow/pull/2413.

With this change, the container properties of a job definition registered by Nextflow in AWS Batch are set starting from the container options defined in the Nextflow task.

**Example**

Given the following container options:
```
process sayHelloFirst {

  containerOptions '--env MYVAR2=foo2 --env MYVAR3=foo3 --memory-swap 240000 --memory-swappiness 20 --shm-size 16000000 --ulimit nofile=1280:2560 --ulimit nproc=16:32 --privileged'

}
```
the job is registered with the following properties on AWS Batch: 

properties

Do note that:
1. this mapping does not include properties already set by other means in Nextflow (like volumes, mount points, cpus, memory, etc.)
2. not all the docker run options can be mapped into AWS.  See [AWS doc](https://docs.aws.amazon.com/batch/latest/APIReference/API_ContainerProperties.html).

### CODE REVIEW

**Comment by manuelesimi** [comment]
I just pushed a fix for tmpfs mounts. They are now correctly set in the container properties:

tmpfs mounts

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AWSContainerOptionsMapper.groovy
Comment (pditommaso) [inline plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AWSContainerOptionsMapper.groovy:21]: Too weak, the container option values could container blanks and quotes. Also I think the parsing should not be done by this component, which could take instead a Map object

**Review by pditommaso — COMMENTED** [review]
Hi Manuale, this is welcome. Actually, I had gave a try some time ago but gave up when I realised container options should be managed at Job Definition level (what a design mistake ..) 

My main comment is that this is a good opportunity to manage the `containerOption` as a map. Look at my commit that I've just pushed 

https://github.com/nextflow-io/nextflow/commit/5e9c1cce8c7037be05a77f4c68f91966d7d62521

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AWSContainerOptionsMapper.groovy
Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AWSContainerOptionsMapper.groovy:21]: That makes sense. I also agree that the parsing should occur earlier. Let me think about it.

Comment (pditommaso) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AWSContainerOptionsMapper.groovy:21]: There's an implementation in the commit I've linked above

Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AWSContainerOptionsMapper.groovy:21]: I saw it :)

Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AWSContainerOptionsMapper.groovy:21]: I saw it, thanks :)

**Comment by manuelesimi** [comment]
Great work (and fast!).
I noticed you pushed your change for the map options to another branch. How should I proceed? Merger yours in this one, wait until you merge yours into master and then merge master back into this, or else?

**Comment by pditommaso** [comment]
I would say to merge my branch in your PR, then you can continue with that

**Comment by manuelesimi** [comment]
@pditommaso There is a problem with your parser. The Map doesn't consider options that can be repeated multiple times (like --env or --tmpfs). Each option should be mapped to a List of values.

**Comment by manuelesimi** [comment]
Also, '=' cannot be used as separator in the `QuoteStringTokenizer`. There are options that uses '=' in their values. E.g. `--env MY_VAR-value`.

**Comment by pditommaso** [comment]
It's getting complicated. Should repeatable options be supported? my original plan was only to support a subset of them.

**Comment by manuelesimi** [comment]
> It's getting complicated. Should repeatable options be supported? my original plan was only to support a subset of them.

Parsing command line options is never easy.

I made a change that seems to work:
From `'--env VAR_FOO -e VAR_FOO2=value2 --env VAR_FOO3=value3 --user nf-user'`, I get:

list

Would that be OK with you?

What's probably more concerning is that we should support only the format "option value" (which is the one documented in the docker run help), not "option=value". So a value with an equal char is not separated (like in the screenshot above).

**Comment by pditommaso** [comment]
Is basically a `Map<String,List>` ?

**Comment by manuelesimi** [comment]
> Is basically a `Map<String,List>` ?

yes, it is.

**Comment by pditommaso** [comment]
Maybe we should model it as a `List` of pairs, or even a custom holder class

**Comment by manuelesimi** [comment]
I'd go for a custom holder where we can add the lookup methods we will need.

**Comment by pditommaso** [comment]
👍

**Comment by manuelesimi** [comment]
Question:

The method `Map getContainerOptionsMap()` in TaskConfig returns a Map. Hence, we have the following options:

1) we change the method signature to return our custom holder class. As far as I see, it is used only [here](https://github.com/nextflow-io/nextflow/blob/6adfce099ecf3c835e03d1a2eb98ea81d0540019/plugins/nf-google/src/main/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandler.groovy#L349) so it would have a low impact.
2) our custom holder extends a Map (but then we need to cast back to the holder's class to access its methods, which I'd really dislike)
3) we stick to a Map<String,List>

Which one would you favor?

**Comment by pditommaso** [comment]
I would say 1.

**Comment by manuelesimi** [comment]
Also a combination of 1. and 2. would work well.

**Comment by manuelesimi** [comment]
> I would say 1.

Sounds good.

**Comment by pditommaso** [comment]
As you wish 😄

**Comment by manuelesimi** [comment]
@pditommaso I switched to a custom handler for the option map. I also added support for long option with dashes (docker has a few we have to support like `-memory-swap ... --shm-size ...`.
I adapted the AWS container mapper accordingly, which is greatly simplified with respect to the previous version.

See what you think.

PS: I'm marking this PR as draft since I need additional tests on AWS to make sure all is working as expected after these changes.

**Comment by manuelesimi** [comment]
I have tested again the submissions to AWS Batch with the latest changes pushed in this PR. I can confirm that all the container properties are set correctly (see the screenshots from previous comments) according to the options specified in `containerOptions`.

As I mentioned, not all the `docker run` options are supported in the AWS job definition. If this PR is approved, should we also update the documentation?

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (pditommaso) [inline modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:8]: Whenever possible new classes should be annotated with `@CompileStatic`

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy
Comment (pditommaso) [inline plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:17]: Let's add `@CompileStatic`

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (pditommaso) [inline modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:27]: Does this always return a `List`? in this case worth declaring `List getValues(String key)`

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (pditommaso) [reply modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:8]: It's missing `CmdLineOptionMapTest` or I'm wrong?

**Review by pditommaso — CHANGES_REQUESTED** [review]
It looks a nice work! Left a few comments

**Comment by pditommaso** [comment]
> should we also update the documentation?

it would be nice!

**Comment by manuelesimi** [comment]
> > should we also update the documentation?
> 
> it would be nice!

See https://github.com/nextflow-io/nextflow/pull/2471/commits/75c5623c0122c70a6ba431d0cbb140b19cce0102

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy
Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:19]: Done

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (manuelesimi) [reply modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:11]: Indeed, it was missing. Just added.

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (manuelesimi) [reply modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy]: done

**Comment by manuelesimi** [comment]
> It looks a nice work! Left a few comments

Hi @pditommaso, thanks for your additional comments. Indeed they helped to improve. 

After the changes, I've tested again with AWS Batch and the job properties are diligently matched with the container options :)

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy
Comment (pditommaso) [inline plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:132]: Is there a side-effect setting an empty `LinuxParameters` object? 

Also, it may be better to have more functional behaviour having `checkLinuxParameters` returning the `LinuxParameters`. 

Finally `checkLinuxParameters` could be a static method to prevent creating a class for each task.

Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:132]: > Is there a side-effect setting an empty `LinuxParameters` object?
> 

It's null by default. I wanted to keep it that way if there are no parameters set. This way we have the same (super-tested) behavior previous to this PR. If no param is set the object is null.

> Also, it may be better to have more functional behaviour having `checkLinuxParameters` returning the `LinuxParameters`.

It makes sense.
> 
> Finally `checkLinuxParameters` could be a static method to prevent creating a class for each task.

Makes sense as well, I'll change it (eventually, it'd be an object for each task :) )

Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy]: I made the entire AWSContainerMapper class static to avoid creating unnecessary objects for each task.

File: modules/nf-commons/src/main/nextflow/util/CmdLineHelper.groovy
Comment (pditommaso) [inline modules/nf-commons/src/main/nextflow/util/CmdLineHelper.groovy:82]: Think this can be removed.

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (pditommaso) [inline modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:81]: I've added the logic to make it hasheable

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy
Comment (pditommaso) [inline plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:22]: Made it return the `ContainerProperties` object

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy
Comment (pditommaso) [inline plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:132]: It should be added a unit test each of the option

File: plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsBatchTaskHandlerTest.groovy
Comment (pditommaso) [inline plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsBatchTaskHandlerTest.groovy:469]: A few tests were failing

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsBatchTaskHandler.groovy
Comment (pditommaso) [inline plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsBatchTaskHandler.groovy:498]: Add the `containerOptions` to the hashing key to force a new Job definition when container a value change

**Comment by pditommaso** [comment]
Just realised that `AwsContainerOptionsMapper` was tested via indirectly the `TaskHandler`. Think it would be better to simplify this testing directly the class ie. 

```
 when:
 def map = CmdLineHelper.parseGnuArgs('--ulimit nofile=1280:2560 --ulimit nproc=16:32')
 def properties = AwsContainerOptionsMapper.createContainerOpts(map) 
 then:
 properties.getUlimits().size() == 2
 properties.getUlimits().get(0).toString() == '{HardLimit: 2560,Name: nofile,SoftLimit: 1280}'
 properties.getUlimits().get(1).toString() == '{HardLimit: 32,Name: nproc,SoftLimit: 16}'
```


https://github.com/nextflow-io/nextflow/blob/80dbfa8e64e42b49acf5b89040b8d9e180565c28/plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsContainerOptionsMapperTest.groovy#L12-L156

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (manuelesimi) [reply modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy:81]: yep, definitely needed and welcomed.

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy
Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:22]: Looks more readable. Thanks.

File: plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsBatchTaskHandlerTest.groovy
Comment (manuelesimi) [reply plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsBatchTaskHandlerTest.groovy:469]: thanks!

File: modules/nf-commons/src/main/nextflow/util/CmdLineHelper.groovy
Comment (manuelesimi) [reply modules/nf-commons/src/main/nextflow/util/CmdLineHelper.groovy:82]: > Just realised that `AwsContainerOptionsMapper` was tested via indirectly the `TaskHandler`. Think it would be better to simplify this testing directly the class ie.
> 
> ```
>  when:
>  def map = CmdLineHelper.parseGnuArgs('--ulimit nofile=1280:2560 --ulimit nproc=16:32')
>  def properties = AwsContainerOptionsMapper.createContainerOpts(map) 
>  then:
>  properties.getUlimits().size() == 2
>  properties.getUlimits().get(0).toString() == '{HardLimit: 2560,Name: nofile,SoftLimit: 1280}'
>  properties.getUlimits().get(1).toString() == '{HardLimit: 32,Name: nproc,SoftLimit: 16}'
> ```
> 
> https://github.com/nextflow-io/nextflow/blob/80dbfa8e64e42b49acf5b89040b8d9e180565c28/plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsContainerOptionsMapperTest.groovy#L12-L156

It is also tested there. When you asked for a test, I thought you wanted a dedicated test class. I agree and will move the tests  as you suggested.

**Comment by manuelesimi** [comment]
> Just realised that `AwsContainerOptionsMapper` was tested via indirectly the `TaskHandler`. Think it would be better to simplify this testing directly the class ie.
> 
> ```
>  when:
>  def map = CmdLineHelper.parseGnuArgs('--ulimit nofile=1280:2560 --ulimit nproc=16:32')
>  def properties = AwsContainerOptionsMapper.createContainerOpts(map) 
>  then:
>  properties.getUlimits().size() == 2
>  properties.getUlimits().get(0).toString() == '{HardLimit: 2560,Name: nofile,SoftLimit: 1280}'
>  properties.getUlimits().get(1).toString() == '{HardLimit: 32,Name: nproc,SoftLimit: 16}'
> ```
> 
> https://github.com/nextflow-io/nextflow/blob/80dbfa8e64e42b49acf5b89040b8d9e180565c28/plugins/nf-amazon/src/test/nextflow/cloud/aws/batch/AwsContainerOptionsMapperTest.groovy#L12-L156

It is also tested there. When you asked for a test, I thought you wanted a dedicated test class. I agree and will move the tests as you suggested.

**Comment by manuelesimi** [comment]
> > Just realised that `AwsContainerOptionsMapper` was tested via indirectly the `TaskHandler`. Think it would be better to simplify this testing directly the class ie.
> > ```
> >  when:
> >  def map = CmdLineHelper.parseGnuArgs('--ulimit nofile=1280:2560 --ulimit nproc=16:32')
> >  def properties = AwsContainerOptionsMapper.createContainerOpts(map) 
> >  then:
> >  properties.getUlimits().size() == 2
> >  properties.getUlimits().get(0).toString() == '{HardLimit: 2560,Name: nofile,SoftLimit: 1280}'
> >  properties.getUlimits().get(1).toString() == '{HardLimit: 32,Name: nproc,SoftLimit: 16}'
> > ```
> > 
> 
> It is also tested there. When you asked for a test, I thought you wanted a dedicated test class. I agree and will adjust the tests as you suggested.

done, see https://github.com/nextflow-io/nextflow/pull/2471/commits/d1d9da011fb3e422b08ebc9b3352768f25605984

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy
Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy:132]: done, see https://github.com/nextflow-io/nextflow/pull/2471/commits/d1d9da011fb3e422b08ebc9b3352768f25605984

File: modules/nf-commons/src/main/nextflow/util/CmdLineHelper.groovy
Comment (manuelesimi) [reply modules/nf-commons/src/main/nextflow/util/CmdLineHelper.groovy]: done, see https://github.com/nextflow-io/nextflow/pull/2471/commits/a724067fe5c8d34ee7685bdb215b212ce7446b83

**Comment by manuelesimi** [comment]
I think I've addressed your latest comments. Pls, let me know if I missed something. Thanks!

**Comment by manuelesimi** [comment]
Hold on, there is a test failing. I know why. I can fix it later.

**Comment by manuelesimi** [comment]
So, I changed two `GoogleLifeSciencesTaskHandlerTest` tests that uses the new config map. I hope I didn't misinterpret them, but basically I'd expect that when the entrypoint is not set in the map, the default entrypoint is returned. See https://github.com/nextflow-io/nextflow/pull/2471/commits/96c3fd4e79ddfd028380fc0b01aef3494450c59e

File: plugins/nf-google/src/test/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandlerTest.groovy
Comment (pditommaso) [inline plugins/nf-google/src/test/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandlerTest.groovy:201]: Interesting, why before it was null? 😕

Comment (pditommaso) [reply plugins/nf-google/src/test/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandlerTest.groovy:201]: So, the constant is `/bin/bash`

https://github.com/nextflow-io/nextflow/blob/96c3fd4e79ddfd028380fc0b01aef3494450c59e/plugins/nf-google/src/main/nextflow/cloud/google/lifesciences/GoogleLifeSciencesConfig.groovy#L48-L48

However when is null defaults to `bash` 

https://github.com/nextflow-io/nextflow/blob/96c3fd4e79ddfd028380fc0b01aef3494450c59e/plugins/nf-google/src/main/nextflow/cloud/google/lifesciences/GoogleLifeSciencesHelper.groovy#L245-L246

Think, it would be better to stick to the current behaviour to avoid introducing unexpected side effects. I'll do the chanege

Comment (pditommaso) [reply plugins/nf-google/src/test/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandlerTest.groovy:201]: Even more intriguing.. actually when `entrypoint` was not specified it was defaulting to `/bin/bash` 

https://github.com/nextflow-io/nextflow/blob/96c3fd4e79ddfd028380fc0b01aef3494450c59e/plugins/nf-google/src/test/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandlerTest.groovy#L185-L185

Not clear why there was that test checking for `null`

Comment (pditommaso) [reply plugins/nf-google/src/test/nextflow/cloud/google/lifesciences/GoogleLifeSciencesTaskHandlerTest.groovy:201]: I've removed that tests 😆

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (pditommaso) [inline modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy]: Usually, it's used `null` for absences of value

File: plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy
Comment (pditommaso) [inline plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy]: This is implicit with groovy, have a look at https://groovy-lang.org/semantics.html#Groovy-Truth

Comment (manuelesimi) [reply plugins/nf-amazon/src/main/nextflow/cloud/aws/batch/AwsContainerOptionsMapper.groovy]: Yep, sorry, that comes from my Java background :)

File: modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy
Comment (manuelesimi) [reply modules/nf-commons/src/main/nextflow/util/CmdLineOptionMap.groovy]: Got it. Thanks.

**Review by pditommaso — APPROVED** [review]
OK, Sunday PR approval! thanks for this, much appreciated!

**Comment by manuelesimi** [comment]
Terrific! This was much needed for our AWS projects.
Thanks for all the help and patience along .

**Comment by pditommaso** [comment]
Released along with version 21.12.1-edge

**Pull request merged**
